themes/qgrid/source/js/menu.js

```diff
@@ -15,41 +15,234 @@ function toggleVisibility(e) {
 	}
 }
 
-function filter(e) {
+function searchOnLoad() {
+	const active = document.querySelector('.active');
+	const search = getSearch();
+	if (active && search) {
+		pageTextSearch(search);
+		tagsSearch(search);
+		menuItemsSearch(search);
+	}
+}
+
+function searchOnHashChange() {
+	const search = getSearch();
+	updateMenuLinks(search);
+	pageTextSearch(search);
+	tagsSearch(search);
+	menuItemsSearch(search);
+}
+
+function onSearchChange(e) {
 	e.stopImmediatePropagation();
 	const input = e.target || e.srcElement;
-	const filter = input.value.toLowerCase();
+	const search = input.value.toLowerCase();
+	setSearch(search);
+}
+
+function setSearch(search) {
+	if (search) {
+		location.hash = 'search=' + search;
+	}
+	else {
+		location.hash = '';
+	}
+}
+
+function getSearch() {
+	const search = new RegExp(/^#search=/);
+	if (search.test(location.hash)) {
+		return location.hash.substr(search.source.length - 1).replace(/%20/g, ' ').toLowerCase();
+	}
+	return '';
+}
+
+function updateMenuLinks(search) {
 	const nav = document.getElementById('nav');
-	const li = nav.getElementsByTagName('li');
-
-	for (let i = 0; i < li.length; i++) {
-		const a = li[i].getElementsByTagName('a')[0];
-		if (a) {
-			if (a.text.toLowerCase().indexOf(filter) > -1) {
-				li[i].style.display = '';
-			} else {
-				li[i].style.display = 'none';
+	const menuItems = nav.getElementsByTagName('li');
+	for (let searchTarget of menuItems) {
+		const title = searchTarget.querySelector('.title');
+		if (title) {
```

Comment:
is it possible that we won't find title? don't need to add necessary checks, because it says(in current line) that title is not required in menu item, and this is not true from semantic.